tools/python/openlr/quality.py

```diff
@@ -0,0 +1,195 @@
+#!/usr/bin/env python3
+
+import xml.etree.ElementTree as ET
+import operator
+
+from itertools import islice
+from collections import namedtuple
+
+LatLon = namedtuple('LatLon', 'lat, lon')
+
+def distance(x, y):
+    return ((x.lat - y.lat)**2 + (x.lon - y.lon)**2) ** 0.5
+
+def curve_length(c):
+    if len(c) < 2:
+        return 0.0
+    return sum(
+        distance(c[i], c[i - 1])
+        for i in range(1, len(c))
+    )
+
+class InterpolatedFunc:
+    def __init__(self, points):
+        self._points = points
+        length = sum(distance(points[i], points[i - 1])
+                     for i in range(1, len(points)))
+        self._time = [0.0]
+        for i in range(1, len(points)):
+            self._time.append(
+                self._time[i - 1] + distance(points[i], points[i - 1])
+            )
+        for i in range(1, len(self._time)):
+            self._time[i] /= length
+
+    def __call__(self, t):
+        assert 0.0 <= t <= 1.0
+        try:
+            return LatLon(self._interp_lat(t), self._interp_lon(t))
+        except IndexError:
+            print(self._time)
+            raise
+
+    def _calc_time(self, points):
+        length = sum(points)
+
+    def _interp_lat(self, t):
+        l, r = self._get_bounding_poits(t)
+        if l == r:
+            return self._points[l].lat
+        dt = self._time[r] - self._time[l]
+        return (
+            self._points[l].lat +
+            (self._points[r].lat - self._points[l].lat) * (t - self._time[l]) / dt
+        )
+
+    def _interp_lon(self, t):
+        l, r = self._get_bounding_poits(t)
+        if l == r:
+            return self._points[l].lon
+        dt = self._time[r] - self._time[l]
+        return (
+            self._points[l].lon +
+            (self._points[r].lon - self._points[l].lon) * (t - self._time[l]) / dt
+        )
+
+    def _get_bounding_poits(self, t):
+        l, r = 0, len(self._points)
+        while l < r:
+            m = (l + r) // 2
+            if t < self._time[m]:
+                r = m
+            elif t > self._time[m]:
+                l = m + 1
+            else:
+                return m, m
+        return (m - 1, m) if t < self._time[m] else (m, m + 1)
+
+def time_integrated(l1, l2, step=10**-2):
+    t = 0.0
+    result = 0.0
+    l1 = InterpolatedFunc(l1)
+    l2 = InterpolatedFunc(l2)
+    while t <= 1.0:
+        result += distance(l1(t), l2(t))
+        t += step
+    return result
+
+def common_subpath(l1, l2, eq=operator.eq):
+    prefs_len = [
+        [0 for _ in range(len(l2) + 1)]
+        for _ in range(len(l1) + 1)
+    ]
+    for i in range(1, len(l1) + 1):
+        for j in range(1, len(l2) + 1):
+            if eq(l1[i - 1], l2[j - 1]):
+                prefs_len[i][j] = prefs_len[i - 1][j - 1] + 1
+    l, i, j = max((
```

Comment:
Otherwise you need to set prefs_len[i][j] to max(prefs_len[i - 1][j], prefs_len[i][j - 1]), no?